---
title: "SummarizedExperiment Tutorial"
author: Wade Boohar, Kayla Xu
date: 09/24/23
updated: 10/17/24
---

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/home1/yuhanyua/490_cluster/analysis_data")) 
```


```{r Package Setup}
# install, only if needed, and load packages (SummarizedExperiment and TCGAbiolinks)
library(TCGAbiolinks)
library(SummarizedExperiment)
```

Load in the RNAseq data
```{r}
rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

rna_se <- GDCprepare(rna_query)

```


###### (1) Summarized Experiment Basics

Let's take a look at rna_se (RNA Summarized Experiment). Call head() and str(). Also explore the data frame by double clicking on it from the Environment panel (or calling "View(rna_se)").

```{r Exercise 1.1}
# write code here
head(rna_se)
str(rna_se)
```

Use the dim() function to find the dimensions of the colData, rowRanges@elementMetadata, and unstranded dataframes. Record them below:

  ~ colData (rows, cols): 1231 88
  ~ rowRanges@elementMetadata: 60660 10
  ~ unstranded: 60660 1231

```{r Exercise 1.2}
# write code here
dim(rna_se@colData)
dim(rna_se@rowRanges@elementMetadata)
dim(rna_se@assays@data$unstranded)


```

Run the following code and fill out a description of the rows/columns of our data frames. You can use this as a cheat sheet moving forward.
  ~ colData
        - Rows: 
        - Columns: 
  ~ rowRanges@elementMetadata
        - Rows: 
        - Columns:
  ~ unstranded (in assays)
        - Rows: 
        - Columns: 
  
```{r Exercise 1.3}
# colData: 


# rowRanges:

# what are the rows... Hint: Do the rows match a dimension of another dataframe?

# for unstranded, don't call anything, just use your logic from earlier!
```

###### (2) ColData

Using the nested dataframe colData in rna_se, view the first 15 values in the age_at_index column.

```{r Exercise 2.1}
# write code here
rna_se@assays@data$unstranded
head(rna_se@colData$age_at_index,15)
```

Call summary on the age_at_index column. Record the output below:
    Min.: 26
    1st Qu.: 48
    Median: 58
    Mean: 58.29
    3rd Qu.: 67
    Max.: 90
    NA's: 1
    
```{r Exercise 2.2}
# write code here
summary(rna_se@colData$age_at_index,15)
```

Create a data frame called rna_clinical that contains data from colData.

```{r Exercise 2.3}
rna_clinical<-data.frame(rna_se@colData)

```

Use the as.data.frame() function to change this from an S4 object to a data frame object.

```{r Exercise 2.4}
rna_clinical<-as.data.frame(rna_clinical)

```

Three columns in rna_clinical contain nested lists. To save the dataframe as a csv later, we need to first remove those columns. Using boolean masks, subset out treatments, primary_site, and disease_type

```{r Exercise 2.5}
column_mask<-ifelse(colnames(rna_clinical)%in%c("treatments","primary_site","disease_type"),F,T)
rna_clinical<-rna_clinical[column_mask]

write.csv(rna_clinical,"/home1/yuhanyua/490_cluster/analysis_data/BCRA_rna_clinical.csv")


```


###### (3) RowRanges

Call head() on rowRanges@elementMetadata:

```{r Exercise 3.1}
# write code here
head(rna_se@rowRanges@elementMetadata)
rna_se@rowRanges@elementMetadata

```

Create a data frame called rna_genes that contains data from elementMetadata.

```{r Exercise 3.2}
rna_genes<-data.frame(rna_se@rowRanges@elementMetadata)


```

Use the as.data.frame() function to change this from an S4 object to a data frame object

```{r Exercise 3.3}
rna_genes<-as.data.frame(rna_genes)
write.csv(rna_genes,"/home1/yuhanyua/490_cluster/analysis_data/BCRA_rna_genes.csv")

```

###### (4) Assays

What are the names of the 6 assays?

```{r Exercise 4.1}
# access the assays here
assayNames(rna_se)

```

Create a new data frame called "rna_counts" that contains the data from unstranded. Recall the patient with NA age data? We need to exclude that patient. We can use the same mask to do this.

```{r Exercise 4.2}
rna_counts<-data.frame(rna_se@assays@data$unstranded)
rna_counts<-as.data.frame(rna_counts)

```

Fetch the data for the first 5 patients and the first 15 rows.

```{r Exercise 4.3}
rna_counts[1:15,1:5]
```


###### (5) Cleaning up your data frames

Call rownames() and colnames() on rna_clinical.

```{r Exercise 5.1}
rownames(rna_clinical)
colnames(rna_clinical)

```

Just like we did to "clinical", we are going to create a column in rna_clinical for age_category. Our cutoff will be "young" as <= 58 y/o, and "old" as > 58 y/o.

There are three age columns in rna_clinical. We'll use the column age_at_index, because it has the fewest number of NA values (only 1).

Age columns:
  rna_clinical$age_at_diagnosis (16 NAs) (in days, not years)
  rna_clinical$age_at_index (1 NA) 
  rna_clinical$paper_age_at_initial_pathologic_diagnosis (19 NAs)
  

Use the information in age_at_index with an ifelse() statement to create a new column in rna_clinical called age_category. Remember ifelse(expression, x, y), where x happens if the expression is true, and y happens if it is false.

```{r Exercise 5.4}
age_na_mask <- !is.na(rna_clinical$age_at_index)
rna_clinical <- rna_clinical[age_na_mask, ]
rna_clinical$age_category<-ifelse(rna_clinical$age_at_index>58,"old","young")


```

----

Let's move on to rna_genes. Call rownames() and colnames() on rna_genes.

```{r Exercise 5.5}
rownames(rna_genes)
colnames(rna_genes)


```

As you can see, both the row names and column names are informative.
----

Now that rna_genes has informative row/column names, let's move on to rna_counts. Call rownames() and colnames() on rna_counts.

```{r Exercise 5.7}
colnames(rna_counts)
rownames(rna_counts)

```

Uh oh! Looks like neither the row nor column names are very informative. Given what we did to the other data frames, add column and row names to rna_counts. Name the columns based on the patient barcodes (ie the row names for rna_clinical), and name the rows based on the gene id's (ie the column names for rna_gene).

```{r Exercise 5.8}
colnames(rna_counts)<-rna_clinical$patient
rownames(rna_counts)<-rna_genes$gene_id

write.csv(rna_counts,"/home1/yuhanyua/490_cluster/analysis_data/BCRA_rna_counts.csv")
rna_counts<-read.csv("/home1/yuhanyua/490_cluster/analysis_data/BCRA_rna_counts.csv")


```

Now, if we look at the data, we'll see the row names set as the gene IDs and the column names as the patient sample barcodes.
```{r}
rna_genes
rna_clinical
```

----

The last step in cleanup is to subset out any samples that are of a normal tissue type. Call unique() on the definition column of rna_clinical.

```{r Exercise 5.9}
# write code here]
unique(rna_clinical$definition)

```

Notice that there are three sample types: "Solid Tissue Normal", "Primary solid Tumor", "Additional Metastatic", and "Metastatic". For now, we are simply going to get rid of the "Solid Tissue Normal" samples, but keep this in mind for your midterm and final projects as these samples could provide you with a good control!

Create a mask that calls "Solid Tissue Normal" FALSE, and the other two sample types TRUE.

```{r Exercise 5.10}
# write code here
solid_tissue_normal_mask <- rna_clinical$definition !="Solid Tissue Normal"


```

You will need to apply this mask to two of our three data frames (rna_clinical, rna_genes, rna_counts). Think about which df's involve samples, and apply the mask to these data frames to subset out the normal samples. You may overwrite the data frames in this case.

```{r Exercise 5.11}
# write code here
rna_clinical <- rna_clinical[solid_tissue_normal_mask, ]
rna_counts <- rna_counts[ ,solid_tissue_normal_mask]

#patient_mask <- ifelse(colnames(rna_counts) %in% rna_clinical$patient,T,F)
#rna_counts_1 <- rna_counts[, patient_mask]

#cat("Number of columns in rna_counts after subsetting:", ncol(rna_counts_1), "\n")

#unique(rna_clinical$patient)
#unique(colnames(rna_counts))
```

  
Since there still is one patients who has "Not Reported" for vital status in our dataframes, we need to remove the patient's information form both rna_clinical and rna_counts. Use the is.na() function to accomplish this here:

```{r Exercise 5.12}
#not_reported_mask<-ifelse(rna_clinical$vital_status=="Not Reported",F,T)
#rna_clinical<-rna_clinical[not_reported_mask, ]
vital_status_mask <- !is.na(rna_clinical$vital_status) & rna_clinical$vital_status != "Not Reported"
rna_clinical <- rna_clinical[vital_status_mask, ]
matching_columns <- colnames(rna_counts) %in% rna_clinical$patient
rna_counts <- rna_counts[, matching_columns]

#rna_clinical$vital_status

```

Great! Now all of our data frames are ready for exploration and analysis.

  
###### (6) Saving data as csv

Loading data from TCGA can be time consuming. It's much faster to access csv files.

Save rna_clinical to your local machine

```{r Exercise 6.1}
write.csv(rna_clinical, "/home1/yuhanyua/490_cluster/analysis_data/brca_rna_clinical_data.csv", row.names = TRUE)
```


Save rna_genes to your local machine.

```{r Exercise 6.2}
write.csv(rna_genes, "/home1/yuhanyua/490_cluster/analysis_data//brca_rna_gene_data.csv", row.names = FALSE)
```


Save rna_counts to your local machine (warning: this will likely take a while!)

```{r Exercise 6.3}
write.csv(rna_counts, "/home1/yuhanyua/490_cluster/analysis_data//brca_rna_count_data.csv", row.names = TRUE)
```
